### ISSUE TITLE
Proposal: Use Distro's Install Repos for Non-Generic Providers

### ISSUE THREAD

**Issue by clnperez** [docker/machine#1148]
I'm running into an issue where the docker port is blocked by a firewall in a cloud provider's Ubuntu 14.04 locked-down image. I feel that machine should handle such scenarios.

So, two options (that I can think of):

1. Add firewall-specific (uwf for Ubuntu, or firewalld, for Redhat/CentOS/Fedora) logic to check if a port is blocked -- and if so, allow it.
2. Use the installer (e.g. apt-get or yum) instead of curl.

What makes this (IMO) not elligible to be pushed out to the drivers is the fact that I'm seeing this with the an openstack driver provider. With an openstack driver, you can run into a multitude of differences that it might be better for machine to just deal with.

Right now I'm leaning toward (and testing out) 2.

Thoughts?

**Comment by clnperez** [comment]
It looks like 1 is going to have to be the way to go, since, by default , docker listens on the UNIX socket, so the distro wouldn't have added any firewall rules for the TCP socket to the installer. And you have to be listening on the TCP port to communicate with docker-machine.

**Comment by clnperez** [comment]
I created PR #1156 to handle this.

**Comment by clnperez** [comment]
Lots more discussion in PR #1156, and this isn't going to be the way to go either. See comments there for more discussion about getting around potential firewall issues.

**Comment by ehazlett** [comment]
Personally I think this is a bit out of scope for machine.  We may potentially have the ability to access ports, etc in the future but this would be at the provider level.  As far as the base OS, we have a handful of "supported" base operating systems and they are expected to have certain base capabilities (such as sudo, root level access, kernel version, etc).  I think once we start doing things deep into the OS we start venturing into the config management which I really want to stay away from.

That being said, we could do a better job of documenting what is needed / expected for the base operating systems for drivers that offer the ability to use custom images (openstack, etc).

As an alternative, have you tried the base Ubuntu cloud image (https://cloud-images.ubuntu.com/trusty/)?  I believe that should work out of the box.

**Comment by clnperez** [comment]
Closing, as this discussion has been picked up in PR #1156.
